crates/forge_services/src/tools/task_list.rs

```diff
@@ -0,0 +1,687 @@
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema, PartialEq)]
+pub struct Stats {
+    /// Total number of tasks in the list.
+    pub total_tasks: u32,
+    /// Number of completed tasks.
+    pub done_tasks: u32,
+    /// Number of pending tasks.
+    pub pending_tasks: u32,
+    /// Number of in-progress tasks.
+    pub in_progress_tasks: u32,
+}
+
+impl std::fmt::Display for Stats {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "<stats>\n<total_tasks>{}</total_tasks>\n<done_tasks>{}</done_tasks>\n<pending_tasks>{}</pending_tasks>\n<in_progress_tasks>{}</in_progress_tasks>\n</stats>",
+            self.total_tasks, self.done_tasks, self.pending_tasks, self.in_progress_tasks)
+    }
+}
+
+/// TaskList operation result.
+#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
+pub struct TaskListResult {
+    /// Message describing the result of the operation.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub message: Option<String>,
+    /// The task that was affected by the operation, if applicable.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub next_task: Option<Task>,
+    /// Statistics about the task list.
+    pub stats: Stats,
+    /// List of tasks
+    pub tasks: Option<Vec<Task>>,
+}
+
+impl Display for TaskListResult {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut result = String::from("<task_list_result>\n");
+
+        if let Some(message) = &self.message {
+            result.push_str(&format!("<message>{message}</message>\n"));
+        }
+
+        if let Some(task) = &self.next_task {
+            result.push_str("<next_task>\n");
+            result.push_str(&format!("{task}\n"));
+            result.push_str("\n</next_task>\n");
+        }
+
+        result.push_str(&format!("{}\n", self.stats));
+
+        if let Some(tasks) = &self.tasks {
+            if !tasks.is_empty() {
+                result.push_str("<tasks_list>\n");
+                for task in tasks {
+                    result.push_str(&format!("{task}\n"));
+                }
+                result.push_str("</tasks_list>\n");
+            }
+        }
+
+        result.push_str("</task_list_result>");
+
+        write!(f, "{result}")
+    }
+}
+
+/// A stateful task management tool that maintains an ordered list of tasks with
+/// status tracking. Provides operations to add tasks (append/prepend), mark
+/// tasks as in-progress (pop_front/pop_back), complete tasks (mark_done), and
+/// view the current state (list). Automatically identifies the next pending
+/// task when completing items and provides detailed statistics on task status.
+/// Ideal for sequential workflows, project planning, and tracking multi-step
+/// processes.
+#[derive(Debug, ToolDescription)]
+pub struct TaskList<F> {
+    infra: Arc<F>,
+    tasks: Arc<Mutex<Vec<Task>>>,
+    next_id: Arc<Mutex<u32>>,
+}
+
+impl<F: Infrastructure> TaskList<F> {
```

Comment:
Create a separate service that does all the task related work. The tool should be simple  and a dumb wrapper for the agent to use that service.